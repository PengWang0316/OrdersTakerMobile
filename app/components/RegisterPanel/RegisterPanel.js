import React, { Component, Fragment } from 'react';
import { View, Text } from 'react-native';
import { Button, Input, Icon } from 'react-native-elements';
import * as Animatable from 'react-native-animatable';
import { connect } from 'react-redux';

import styles from './Styles';
import { checkUsernameAvailable, registerUser } from '../../actions/UserActions';

const USERNAME_REGEXP = /[^\w\d_@]/g;
const EMAIL_REGEXP = /^[\w\d]+@.+\..+/g;
const USERNAME_TIP_MESSAGE = 'Characters, number, _ and @';
const USERNAME_NOAVALIABLE_MESSAGE = 'The user name laready exsits';
const PASSWORD_TIP_MESSAGE = 'Passwords should be same';
const EMAIL_TIP_MESSAGE = 'Wrong email address';

type States = {
  username: string,
  password: string,
  repeatPassword: string,
  email: string,
  isSubmitted: boolean,
  isReady: boolean,
  isChecking: boolean,
  isNameAvalible: boolean,
  usernameErrorMessage: boolean,
  passwordErrorMessage: boolean,
  emailErrorMessage: boolean,
  isWaiting: boolean,
};

/**
 * The Register panel component
 */
export class RegisterPanel extends Component<null, States> {
  state = {
    username: '',
    password: '',
    repeatPassword: '',
    email: '',
    isSubmitted: false,
    isReady: false,
    usernameErrorMessage: false,
    passwordErrorMessage: false,
    emailErrorMessage: false,
    isChecking: false,
    isNameAvalible: false,
    isWaiting: false,
  };

  /**
   * Set username state and call checkAllState function.
   * @param {string} target is the target's state name.
   * @param {string} value is the target's text value.
   * @return {null} No return.
   */
  handleInputTextChange = (target, value) => this.setState({ [target]: value }, () => {

    // Do the validation
    switch (target) {
      case 'username':
        this.setState({ usernameErrorMessage: false, isReady: false, isNameAvalible: false }); // Erasing the error state and check it again. If username is changed, alway set the isReady to false and let the checking function callback to decide it state.
        if (value.match(USERNAME_REGEXP)) {
          this.setState({ usernameErrorMessage: USERNAME_TIP_MESSAGE });
        } else {
          // Setting up a timeout function to check the avalibility for the username.
          if (this.checkUsernameTimeout) clearTimeout(this.checkUsernameTimeout); // If the check function has already been schulded, clear the old one.
          this.checkUsernameTimeout = setTimeout(() => {
            this.setState({ isChecking: true });
            checkUsernameAvailable(value).then(isAvalible => this.setState(({
              username, password, repeatPassword, email,
            }) => ({
              isReady: !!(username && password && repeatPassword && email && !!isAvalible),
              isChecking: false,
              isNameAvalible: !!isAvalible,
              usernameErrorMessage: isAvalible ? false : USERNAME_NOAVALIABLE_MESSAGE,
            })));
          }, 1000);
        }
        break;
      case 'password':
        this.setState({ passwordErrorMessage: false }); // Erasing the error state and check it again.
        if (value !== this.state.repeatPassword) {
          this.setState({ passwordErrorMessage: PASSWORD_TIP_MESSAGE, isReady: false });
        }
        break;
      case 'repeatPassword':
        this.setState({ passwordErrorMessage: false }); // Erasing the error state and check it again.
        if (value !== this.state.password) {
          this.setState({ passwordErrorMessage: PASSWORD_TIP_MESSAGE, isReady: false });
        }
        break;
      case 'email':
        this.setState({ emailErrorMessage: false }); // Erasing the error state and check it again.
        if (!value.match(EMAIL_REGEXP)) {
          this.setState({ emailErrorMessage: EMAIL_TIP_MESSAGE, isReady: false });
        }
        break;
      default:
        break;
    }
    // If all field is not empty and no error message shows up, set the isReady to true.
    this.setState(({
      username, password, repeatPassword, email, usernameErrorMessage, passwordErrorMessage, emailErrorMessage,
    }) => {
      return ({ isReady: !!(username && password && repeatPassword && email) && !(usernameErrorMessage || passwordErrorMessage || emailErrorMessage) });
    });
  });

  /**
   * Disable the register button, set isWaiting to true, and call the registerUser function in UserActions.
   * @return {null} No return.
   */
  handleRegisterBtnClick = () => {
    this.setState({
      isReady: false,
      isWaiting: true,
    });
    this.props.registerUser({
      username: this.state.username,
      password: this.state.password,
      email: this.state.email,
    });
  };

  /**
   * THe render method
   * @return {jsx} Return jsx for the component.
   */
  render() {
    const {
      username, password, repeatPassword, email, isSubmitted, isReady, isChecking, isWaiting,
      usernameErrorMessage, passwordErrorMessage, emailErrorMessage, isNameAvalible,
    } = this.state;
    return (
      <View style={styles.internalContainer}>
        <View>
          <Text style={styles.title}>Username Login</Text>
        </View>

        <View style={styles.inputView}>
          <Input
            onChangeText={text => this.handleInputTextChange('username', text)}
            editable={!isWaiting}
            placeholder="Username"
            value={username}
            leftIcon={(
              <Icon
                type="font-awesome"
                name="user-o"
                size={18}
                color="gray"
              />
            )}
            rightIcon={(
              <Fragment>
                {isChecking && (
                  <Animatable.View
                    animation="rotate"
                    easing="linear"
                    iterationCount="infinite"
                  >
                    <Icon
                      type="evilicon"
                      name="spinner-3"
                      size={18}
                      color="gray"
                    />
                  </Animatable.View>
                )}
                {isNameAvalible && <Icon name="check-circle" color="green" />}
              </Fragment>
            )}
            errorMessage={usernameErrorMessage}
          />
        </View>

        <View style={styles.inputView}>
          <Input
            onChangeText={text => this.handleInputTextChange('password', text)}
            editable={!isWaiting}
            value={password}
            secureTextEntry
            placeholder="Password"
            leftIcon={(
              <Icon
                type="feather"
                name="lock"
                size={18}
                color="gray"
              />
            )}
          />
        </View>

        <View style={styles.inputView}>
          <Input
            onChangeText={text => this.handleInputTextChange('repeatPassword', text)}
            editable={!isWaiting}
            value={repeatPassword}
            secureTextEntry
            placeholder="Repeat Password"
            leftIcon={(
              <Icon
                name="repeat"
                size={18}
                color="gray"
              />
            )}
            errorMessage={passwordErrorMessage}
          />
        </View>

        <View style={styles.inputView}>
          <Input
            onChangeText={text => this.handleInputTextChange('email', text)}
            editable={!isWaiting}
            value={email}
            placeholder="Email"
            leftIcon={(
              <Icon
                type="material-icons"
                name="mail-outline"
                size={18}
                color="gray"
              />
            )}
            errorMessage={emailErrorMessage}
          />
        </View>

        <View style={styles.buttonView}>
          <Button
            disabled={isSubmitted || !isReady}
            title="LOGIN"
            onPress={this.handleRegisterBtnClick}
            buttonStyle={styles.registerBtn}
          />
        </View>
      </View>
    );
  }
}
/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  registerUser: user => dispatch(registerUser(user)),
});
export default connect(null, mapDispatchToProps)(RegisterPanel);
