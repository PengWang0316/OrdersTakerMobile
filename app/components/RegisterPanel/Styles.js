import { StyleSheet } from 'react-native';

import Theme from '../../Theme';

export default StyleSheet.create({
  internalContainer: {
    width: '75%',
  },
  inputView: {
    marginBottom: 20,
  },
  title: {
    fontWeight: 'bold',
    fontSize: 20,
    marginBottom: 15,
    color: Theme.primary.main,
  },
  buttonView: {
    marginTop: 30,
  },
  registerBtn: {
    backgroundColor: Theme.primary.main,
    borderRadius: 15,
  },
});
