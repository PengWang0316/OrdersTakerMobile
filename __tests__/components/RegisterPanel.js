import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import { RegisterPanel } from '../../app/components/RegisterPanel/RegisterPanel';

jest.mock('../../app/actions/UserActions', () => ({ checkUsernameAvailable: jest.fn().mockReturnValue(Promise.resolve(true)) }));

describe('RegisterPanel', () => {
  const defaultProps = {
    registerUser: jest.fn(),
    navigation: { goBack: jest.fn() },
  };
  const getShallowComponent = (props = defaultProps) => shallow(<RegisterPanel {...props} />);

  test('Initial states', () => {
    const component = getShallowComponent();
    const {
      username,
      password,
      repeatPassword,
      email,
      isSubmitted,
      isReady,
      isChecking,
      isNameAvalible,
      usernameErrorMessage,
      passwordErrorMessage,
      emailErrorMessage,
      isWaiting,
    } = component.state();

    expect(username).toBe('');
    expect(password).toBe('');
    expect(repeatPassword).toBe('');
    expect(email).toBe('');
    expect(isSubmitted).toBe(false);
    expect(isReady).toBe(false);
    expect(isChecking).toBe(false);
    expect(isNameAvalible).toBe(false);
    expect(usernameErrorMessage).toBe(false);
    expect(passwordErrorMessage).toBe(false);
    expect(emailErrorMessage).toBe(false);
    expect(isWaiting).toBe(false);
  });

  test('handleRegisterBtnClick', () => {
    const component = getShallowComponent();
    component.setState({
      isReady: true, isWaiting: false, username: 'username', password: 'password', email: 'email',
    });
    component.instance().handleRegisterBtnClick();

    expect(component.state('isReady')).toBe(false);
    expect(component.state('isWaiting')).toBe(true);
    expect(defaultProps.registerUser).toHaveBeenCalledTimes(1);
    expect(defaultProps.registerUser).toHaveBeenLastCalledWith({
      username: 'username', password: 'password', email: 'email',
    });
    expect(defaultProps.navigation.goBack).toHaveBeenCalledTimes(1);
  });

  test('handleInputTextChange target username match USERNAME_REGEXP', () => {
    const component = getShallowComponent();
    component.setState({ usernameErrorMessage: true, isReady: true, isNameAvalible: true });
    component.find('Input').at(0).simulate('changeText', 'df#df');
    // component.instance().handleInputTextChange('username', 'df#df');

    expect(component.state('username')).toBe('df#df');
    expect(component.state('usernameErrorMessage')).toBe('Characters, number, _ and @');
    expect(component.state('isReady')).toBe(false);
    expect(component.state('isNameAvalible')).toBe(false);
  });

  test('handleInputTextChange target username does not match USERNAME_REGEXP', () => {
    jest.useFakeTimers();
    const UserActions = require('../../app/actions/UserActions');
    clearTimeout = jest.fn();
    const component = getShallowComponent();
    component.setState({
      usernameErrorMessage: true, isReady: false, isNameAvalible: false, isChecking: false, password: 'pw', repeatPassword: 'repw', email: 'email',
    });
    component.instance().handleInputTextChange('username', 'dfdf124_df@');

    jest.runAllTimers();
    expect(clearTimeout).not.toHaveBeenCalled();
    expect(component.state('username')).toBe('dfdf124_df@');
    expect(component.state('usernameErrorMessage')).toBe(false);
    expect(component.state('isReady')).toBe(true);
    // expect(UserActions.checkUsernameAvailable).toHaveBeenCalledTimes(1); jest.mock is not hoisted automatically. So, I cannot test this.
    component.instance().handleInputTextChange('username', 'dfdf124_df@');
    expect(clearTimeout).toHaveBeenCalledTimes(1);
  });

  test('handleInputTextChange target password value equal repeatPassword', () => {
    const component = getShallowComponent();
    component.setState({ repeatPassword: 'password', passwordErrorMessage: false, isReady: true });
    component.find('Input').at(1).simulate('changeText', 'password');
    // component.instance().handleInputTextChange('password', 'password');

    expect(component.state('password')).toBe('password');
    expect(component.state('passwordErrorMessage')).toBe(false);
    expect(component.state('isReady')).toBe(false);
  });

  test('handleInputTextChange target password value not equal repeatPassword', () => {
    const component = getShallowComponent();
    component.setState({ repeatPassword: 'reassword', passwordErrorMessage: false, isReady: true });
    component.instance().handleInputTextChange('password', 'password');

    expect(component.state('password')).toBe('password');
    expect(component.state('passwordErrorMessage')).toBe('Passwords should be same');
    expect(component.state('isReady')).toBe(false);
  });

  test('handleInputTextChange target repeatPassword value equal password', () => {
    const component = getShallowComponent();
    component.setState({ password: 'repeatPassword', passwordErrorMessage: false, isReady: true });
    component.find('Input').at(2).simulate('changeText', 'repeatPassword');
    // component.instance().handleInputTextChange('repeatPassword', 'repeatPassword');

    expect(component.state('repeatPassword')).toBe('repeatPassword');
    expect(component.state('passwordErrorMessage')).toBe(false);
    expect(component.state('isReady')).toBe(false);
  });

  test('handleInputTextChange target repeatPassword value not equal password', () => {
    const component = getShallowComponent();
    component.setState({ password: 'peassword', passwordErrorMessage: false, isReady: true });
    component.instance().handleInputTextChange('repeatPassword', 'password');

    expect(component.state('repeatPassword')).toBe('password');
    expect(component.state('passwordErrorMessage')).toBe('Passwords should be same');
    expect(component.state('isReady')).toBe(false);
  });

  test('handleInputTextChange target email match EMAIL_REGEXP', () => {
    const component = getShallowComponent();
    component.setState({ emailErrorMessage: false });
    component.find('Input').at(3).simulate('changeText', 'df@df.com');
    // component.instance().handleInputTextChange('email', 'df@df.com');

    expect(component.state('email')).toBe('df@df.com');
    expect(component.state('emailErrorMessage')).toBe(false);
  });

  test('handleInputTextChange target email not match EMAIL_REGEXP', () => {
    const component = getShallowComponent();
    component.setState({ emailErrorMessage: false });
    component.instance().handleInputTextChange('email', 'df@dfcom');

    expect(component.state('email')).toBe('df@dfcom');
    expect(component.state('emailErrorMessage')).toBe('Wrong email address');
    component.instance().handleInputTextChange('', 'df@dfcom'); // for the switch default
  });

  test('Snapshot', () => expect(renderer.create(<RegisterPanel />).toJSON()).toMatchSnapshot());
});
