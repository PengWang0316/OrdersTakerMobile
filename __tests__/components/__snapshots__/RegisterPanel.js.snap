// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterPanel Snapshot 1`] = `
<View>
  <Text>
    Register Panel
  </Text>
</View>
`;
